src/main/scala/org/opennetworkinsight/netflow/FlowPostLDA.scala

```diff
@@ -63,67 +65,51 @@ object FlowPostLDA {
+    val dataWithDestScore = score(sc, dataWithSrcScore, docToTopicMixDF, words, DestinationScore, DestinationIP, DestinationProbabilities, DestinationWord)
+    val dataScored = minimumScore(dataWithDestScore)
 
-    // Using dropRight as we don't need last column MinimumScore, only SourceScore and DestinationScore
-    outputRDD.map(row => Row.fromSeq(row.toSeq.dropRight(1))).map(_.mkString(outputDelimiter)).saveAsTextFile(resultsFilePath)
+    logger.info("Persisting data")
+    val filteredDF = dataScored.filter(MinimumScore + " < " + threshold)
+    filteredDF.orderBy(MinimumScore).limit(topK).rdd.map(row => Row.fromSeq(row.toSeq.dropRight(1))).map(_.mkString(outputDelimiter)).saveAsTextFile(resultsFilePath)
 
     logger.info("Flow post LDA completed")
-
   }
 
   def score(sc: SparkContext,
             dataFrame: DataFrame,
-            ipToTopicMixes: Map[String, Array[Double]],
-            wordToProbPerTopic: Map[String, Array[Double]],
+            docToTopicMixesDF: DataFrame,
+            words: Broadcast[Map[String, Array[Double]]],
             newColumnName: String,
             ipColumnName: String,
+            ipProbabilitiesColumnName: String,
             wordColumnName: String): DataFrame = {
 
-    val topics = sc.broadcast(ipToTopicMixes)
-    val words = sc.broadcast(wordToProbPerTopic)
+    val dataWithIpProbJoin = dataFrame.join(docToTopicMixesDF, dataFrame(ipColumnName) === docToTopicMixesDF(Doc))
 
-    def scoreFunction(ip: String, word: String): Double = {
-      val uniformProb = Array.fill(20) {
-        0.05d
-      }
+    var newSchemaColumns = dataFrame.schema.fieldNames :+ Probabilities + " as " + ipProbabilitiesColumnName
+    val dataWithIpProb = dataWithIpProbJoin.selectExpr(newSchemaColumns: _*)
 
-      val topicGivenDocProbs = topics.value.getOrElse(ip, uniformProb)
-      val wordGivenTopicProbs = words.value.getOrElse(word, uniformProb)
+    def scoreFunction(word: String, ipProbabilities: Seq[Double]): Double = {
+      val uniformProb = Array.fill(20)(0.05d)
+      val wordGivenTopicProb = words.value.getOrElse(word, uniformProb)
 
-      topicGivenDocProbs.zip(wordGivenTopicProbs)
+      ipProbabilities.zip(wordGivenTopicProb)
         .map({ case (pWordGivenTopic, pTopicGivenDoc) => pWordGivenTopic * pTopicGivenDoc })
         .sum
     }
 
-    def udfScoreFunction = udf((ip: String, word: String) => scoreFunction(ip, word))
+    def udfScoreFunction = udf((word: String, ipProbabilities: Seq[Double]) => scoreFunction(word, ipProbabilities))
 
-    dataFrame.withColumn(newColumnName, udfScoreFunction(dataFrame(ipColumnName), dataFrame(wordColumnName)))
+    val result: DataFrame = dataWithIpProb.withColumn(newColumnName, udfScoreFunction(dataWithIpProb(wordColumnName), dataWithIpProb(ipProbabilitiesColumnName)))
+    newSchemaColumns = dataFrame.schema.fieldNames :+ newColumnName
+    result.select(newSchemaColumns.map(col): _*)
```

Comment:
First when we join the _dataFrame_ with the documents per topic data frame we add the 20 probabilities as a new column. Then _result_ is the data frame with 20 probabilities and a new column called _scoreColumnName_, that means we now have 2 new columns.

The very last step takes the original schema without the 20 probabilities column and just add the score column and do a select of that removing the 20 probabilities column.